### DISCUSSION TITLE
Shadow as an evaluation framework for multiple PETs

### DISCUSSION THREAD

**Discussion by akshayeshenoi** [shadow/shadow#1244] [Q&A] [answered]
Hello! 

Thanks for all the great work! 

We are currently researching the effects of various Privacy Enhancing Technologies (PETs) on the traffic output at the last mile. For now, we are considering replaying a pcap file through various PETs like Tor, VPN and a custom solution.

I have the following questions:
1.  Since we need to replay traffic (both TCP and UDP), we need a plugin that can do that. I noticed the [pcap_replay](https://github.com/shadow/shadow-plugin-extras/tree/master/pcap_replay) plugin hosted in the shadow-plugin-extras repository, but I'm curious to know if it is still functional with the latest release of shadow (v1.14). Alternatively, the [tcpreplay](https://github.com/appneta/tcpreplay) tool by AppNeta seems to be part of a rich ecosystem of other network tools that can be modified to work with Shadow. Note that we also wish to replay UDP traffic. What are your thoughts?

2. In the case of a VPN, do you have any ideas on how we could acheive this on shadow? According to my understanding, tools like OpenVPN create a virutal network device on the client to intercept all application traffic and then relay it to the server counterpart. How feasible would this be on Shadow?

Please let me know your thoughts! 
Thank you for your time.

Regards,
Akshaye

**Comment by robgjansen** [comment, marked as answer]
Sounds like an interesting project!

First of all, some bacakground:
- Shadow _simulates_ TCP and UDP - the internal TCP and UDP implementations are not inter-operable with standard implementations in the wild (i.e., on Linux). Therefore, you will not be able to take packets generated by Shadow and directly send them into a real network, and you will not be able to take packets from a real network and send them directly into Shadow.
- However, you _should_ be able to take packets from a real network, extract the payloads and timing information, and replay the payloads into new TCP and UDP sockets when running in Shadow, and Shadow will encapsulate those payloads into the internal protocols. I believe this is essentially what the `pcap_replay` plugin is doing.

Now to answer your questions:
1. I'm not sure if the `pcap_replay` plugin is still functional. Browsing the code, it appears that it should still work in the latest (as of writing) main branch of Shadow (in the v1.14.X series). I have not tested it though. I would guess that it would not take much to make it work. It looks like it only supports TCP currently (`SOCK_STREAM`) but it could be extended to support UDP (`SOCK_DGRAM`).
2. Shadow does not currently emulate virtual network devices, so I don't think this approach would work in Shadow. You could update Shadow to support it, but I'm not familiar enough with this particular kernel API to properly judge how difficult this would be. I suspect it might be too much work. I wonder if there is a VPN that you could use that operates in application space - something like a SOCKS proxy that acts as a VPN client? It might be easier to take that approach because using standard TCP and UDP sockets (rather than specific kernel API features) is much more likely to work correctly in Shadow.

**Reply by akshayeshenoi** [reply]
Thanks for the detailed explanation!

I managed to get the plugin to work with a few [fixes](https://github.com/shadow/shadow-plugin-extras/pull/9), and will probably try to extend it to support UDP.

Let me see if I find some application level VPN that we can work with on Shadow.
